### DISCUSSION TITLE
TE2/3? Calc Group issue with ordinal?

### DISCUSSION THREAD

**Discussion by TinyTimothy2** [TabularEditor/TabularEditor3#1328] [🙋♀️TE3 Q&A]
Not sure if this applies to TE3.  In TE2 there is no way to have ordinal of -1. (which i thought was the default).  It is buggy once you try to reorder them.  You can also have the same ordinal number for each item and load it. Which of course crashes the visual.

I have also seen where when using SSAS and TE2, if you try to select a FIELD measure (say fx for TITLE) it will duplicate the measure name!!
As far as i can tell this happens when TE2 with SSAS and not PBI directly

**Comment by TinyTimothy2** [comment]
This is what it looks like in Power BI when selecting measure to use with formatting AFTER creating CG with TE2 in SSAS.
Even though only 1 measure is actually created in TE2 and only 1 measure shows in TOM in power bi, you will see 2 measures when you try to use the measure in formatting scenarios
![image](https://github.com/TabularEditor/TabularEditor3/assets/48966375/26d778e1-fcba-444d-b747-9b096853b710)

**Reply by otykier** [reply]
This is a PBI issue that’s unrelated to Tabular Editor (you would see the same thing if you created the calc group in SSMS or VS)

**Reply by TinyTimothy2** [reply]
thank you.  is the cause calculation group? does it happen if you create in directly within PB (now that you can do that in pb?)
Does it matter which you pick? i assume no, i have not noticed a diff.
Does not having a -1 ordinal matter?  I thought maybe it did since it is by default.
I think one thing happening is if you copy a CG with C Items from one model to another it sets them all to 1.  It seems there should be a check in place that all the items have a distinct ordinal.  Maybe TE3 has that

**Reply by edhans** [reply]
The service defines the ordinal as a zero based value. The -1 is what it assigns when you haven't assigned anything, and if you just have 1 item, it doesn't matter. 

Once you have 2+ calc items, assign the proper ordinal values you need, and don't use -1. Just use 0 or positive integers.

**Reply by TinyTimothy2** [reply]
great thanks.

**Comment by TinyTimothy2** [comment]
In TE2 you can still publish with calc items with the same ordinals.  It would be nice if there was a warning, or auto increment.
I DUPLICATE alot, and it dups with same ordinal.  Seems petty, but after 100s it does get tedious

**Reply by otykier** [reply]
FYI, TE2 assigns a new ordinal if you use Ctrl+C/Ctrl+V rather than **right-click > Duplicate** (**Duplicate** by design retains all properties exactly as they are, with the exception of the _Name_ property). If you want to make sure that you don't have two calc items with the same ordinal, you could create a Best Practice Rule using the following expression:

```
Parent.CalculationItems.Any(
	it <> current and
	it.Parent = current.Parent and
	it.Ordinal >= 0 and
	it.Ordinal = current.Ordinal
)
```

![image](https://github.com/TabularEditor/TabularEditor3/assets/8976200/f5b7f803-134d-45a1-a5d8-4d5e87b6bc21)

Once this rule has been created, the Best Practice Analyzer will show you any items within the same calc group, that have the same ordinal:

![image](https://github.com/TabularEditor/TabularEditor3/assets/8976200/f1cfbdc3-b3be-4a30-8679-fa94d7381eb9)
